### PR TITLE
Fix packed struct layout regression (issue #2346)

### PR DESCRIPTION
**Pull request by kinke**
I probably introduced this regression lately. Back then, I didn't account for container aggregates, which require that the LL struct is packed if the overall D aggregate alignment is lower than a field's natural alignment.

### CODE REVIEW

File: tests/codegen/gh2346.d [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+static assert(PackedContainer2.alignof == 1);
+static assert(PackedContainer2.sizeof == 5);
+static assert(PackedContainer2.two.offsetof == 1);
+
+// CHECK-DAG: %gh2346.WeirdContainer = type { i8, [1 x i8], %gh2346.UnalignedUInt, [2 x i8] }
+align(4) struct WeirdContainer {
+    ubyte one;
+    align(2) UnalignedUInt two;
+}
+static assert(WeirdContainer.alignof == 2);
```
Comment (kinke) [inline tests/codegen/gh2346.d]: Hmm, that was a bit surprising.

Comment (kinke) [reply tests/codegen/gh2346.d]: Trying to fix this upstream, see https://github.com/dlang/dmd/pull/7164.

Comment (JohanEngelen) [reply tests/codegen/gh2346.d]: Nice catch.

Comment (JohanEngelen) [reply tests/codegen/gh2346.d]: Comment-out the `static assert(...=4)` then?

File: ir/irtypeaggr.cpp [hunk 2]
```diff
@@ -213,10 +217,12 @@ bool IrTypeAggr::isPacked(AggregateDeclaration *ad) {
     // aligned, and LLVM would insert padding.
     const auto naturalFieldAlignment = field->type->alignsize();
     const auto mask = naturalFieldAlignment - 1;
-
-    // If the aggregate's size is unknown, any field with natural alignment > 1
-    // will make it packed.
-    if ((aggregateSize & mask) != 0 || (field->offset & mask) != 0) {
+    // Additionally, the aggregate's overall alignment must not be lower than
+    // a field's natural one, otherwise LLVM would use a higher overall LL
+    // struct alignment, causing additional padding in aggregates containing
```
Comment (JohanEngelen) [inline ir/irtypeaggr.cpp]: This is related to
```
align(2) struct S {
   align(4) int i;
}
```
?
(perhaps add a test for that too?)

File: tests/codegen/gh2346.d (hunk 1 above)
Comment (kinke) [reply tests/codegen/gh2346.d]: I'll just wait for upstream to be merged and will then embed it here.

Comment (JohanEngelen) [reply tests/codegen/gh2346.d]: nice

File: ir/irtypeaggr.cpp (hunk 2 above)
Comment (kinke) [reply ir/irtypeaggr.cpp]: Hmm you're right, explicit alignment needs to be taken into consideration here as well.

**Comment by kinke** [comment]
Upstream fix (merged in the meantime) incorporated.

**Pull request merged**